### DISCUSSION TITLE
How to draw on especific widget?

### DISCUSSION THREAD

**Discussion by jcbritobr** [fltk-rs/fltk-rs#977] [General]
Hello, good morning.

I'm reading some documents about drawing on surfaces, and I'm having some issues in drawings. When I use the functions widget.draw(...), why it doesn't paint only on widget that called the function? The code below is painting over all surface window.

```rust

use fltk::{
    app,
    draw::draw_rect_fill,
    enums::Color,
    frame::Frame,
    prelude::{GroupExt, WidgetBase, WidgetExt},
    window::Window,
};

fn main() {
    let app = app::App::default();
    let mut win = Window::default().with_size(400, 300).with_label("Painter");

    let mut frame = Frame::default().with_size(300, 200).center_of(&win);
    frame.set_color(Color::White);
    frame.set_frame(fltk::enums::FrameType::DownBox);

    win.end();
    win.show_with_env_args();

    frame.draw(move |f| {
        draw_rect_fill(100, 100, 400, f.height() - 100, Color::Red);
    });

    app.run().expect("Cant execute application.");
}
```

![image](https://user-images.githubusercontent.com/6579862/140502259-65a71e9a-f93f-46e8-8843-1b8ab250202e.png)

**Comment by MoAlyousef** [comment]
Hi
FLTK enables drawing outside the widget's border, that's useful for shadows and labels etc. 
Your options to draw within the widget itself only, is to either:
1- Use coordinates relative to the widget:
```rust
    frame.draw(move |f| {
        draw_rect_fill(f.x() + 50, f.y() + 50, f.w() - 100, f.height() - 100, Color::Red);
    });
```

2- Use push_clip and pop_clip to not worry about anything spilling outside the border of the widget:
```rust
    frame.draw(move |f| {
        push_clip(f.x(), f.y(), f.w(), f.h());
        draw_rect_fill(f.x() + 50, f.y() + 50, f.w(), f.height() - 100, Color::Red);
        pop_clip();
    });
```

3 - Use an Offscreen (which allows 0-based origin drawing (however the Offscreen::copy() method will still require coordinates relative to the frame when copying to it):
```rust
    let ofs = Offscreen::new(frame.w(), frame.h()).unwrap();
    frame.draw(move |f| {
        if ofs.is_valid() {
            ofs.copy(f.x() + 50, f.y() + 50, f.w() - 100, f.h() - 100, 0, 0);
        } else {
            ofs.begin();
            draw_rect_fill(0, 0, f.w(), f.h(), Color::Red);
            ofs.copy(f.x() + 50, f.y() + 50, f.w() - 100, f.h() - 100, 0, 0);
            ofs.end();
        }
    });
```

**Reply by jcbritobr** [reply]
Hello. Thank you. I clearly understood now. 💯
